src/main/java/org/radarcns/service/DataSetService.java

```diff
@@ -197,4 +242,71 @@ private Header getHeader(String project, String subject, String source,
+     * updated with the the relevant sourceDataType of sourceData requested.
+     *
+     * @param projectName of project
+     * @param subjectId of subject
+     * @param sources requested
+     * @param timeWindow interval
+     * @param start time
+     * @param end time
+     * @return calculated data.
+     */
+    public AggregatedDataPoints getAggregatedData(String projectName, String subjectId,
+            List<AggregateDataSource> sources, TimeWindow timeWindow, Instant start, Instant end) {
+        List<DataItem> dataItems = calculateIntervals(start, end, timeWindow).stream()
+                .map(p -> computeAggregatedDataItem(projectName, subjectId, sources, p, timeWindow))
+                .collect(Collectors.toList());
+        // fill up sourceData.type field
+        try {
+            for (AggregateDataSource source : sources) {
+                for (SourceData sourceData : source.getSourceData()) {
+                    SourceDataDTO definition = sourceCatalog.getSourceData(sourceData.getName());
+                    sourceData.setType(definition.getSourceDataType());
+                }
+            }
+        } catch (IOException exe) {
+            throw new BadGatewayException(exe);
+        }
+
+        return new AggregatedDataPoints(projectName, subjectId,
+                dataItems.stream().map(p -> (Integer) p.getValue()).reduce(Integer::max).get(),
+                new TimeFrame(start, end), timeWindow, sources, dataItems);
+    }
+
+    private DataItem computeAggregatedDataItem(String projectName, String subjectId,
+            List<AggregateDataSource> aggregateDataSources, TimeFrame timeFrame,
+            TimeWindow timeWindow) {
+        Integer count = 0;
+        for (AggregateDataSource aggregate : aggregateDataSources) {
+            for (SourceData sourceDataName : aggregate.getSourceData()) {
+                MongoSourceDataWrapper sourceDataWrapper = mongoSensorMap
+                        .get(sourceDataName.getName());
+                if (sourceDataWrapper
+                        .anyRecordsExist(projectName, subjectId, aggregate.getSourceId(),
+                                Date.from(timeFrame.getStartDateTime()),
+                                Date.from(timeFrame.getEndDateTime()), MongoHelper
+                                        .getCollection(mongoClient,
+                                                sourceDataWrapper.getCollectionName(timeWindow)))) {
+                    count++;
+                }
+            }
+        }
+        return new DataItem(count, timeFrame.getStartDateTime());
+    }
+
+    private List<TimeFrame> calculateIntervals(Instant start, Instant end, TimeWindow timeWindow) {
+        List<TimeFrame> timeFrames = new ArrayList<>();
+        TemporalAmount window = RadarConverter.getDuration(timeWindow);
+        TimeFrame timeFrame = new TimeFrame(start, start.plus(window));
+        while (!timeFrame.getEndDateTime().isAfter(end)) {
```

Comment:
Should we add a maximum size to timeFrames here as well? Otherwise scripts may request millions of time frames in a single request.